store/store.go

```diff
@@ -1191,6 +1212,116 @@ func (s *Store) Find(ctx context.Context, search *Search, user *auth.UserState)
 		return nil, ErrBadQuery
 	}
 
+	q := url.Values{}
+	q.Set("fields", strings.Join(s.findFields, ","))
+	q.Set("architecture", s.architecture)
+
+	if search.Private {
+		q.Set("private", "true")
+	}
+
+	if search.Prefix {
+		q.Set("name", searchTerm)
+	} else {
+		if search.CommonID != "" {
+			q.Set("common-id", search.CommonID)
+		}
+		if searchTerm != "" {
+			q.Set("q", searchTerm)
+		}
+	}
+
+	// section is "category" in search v2
+	if search.Section != "" {
+		q.Set("category", search.Section)
+	}
+
+	// with search v2 all risks are searched by default (same as scope=wide
+	// with v1) so we need to restrict channel if scope is not passed.
+	if search.Scope == "" {
+		q.Set("channel", "stable")
+	} else if search.Scope != "wide" {
+		return nil, ErrInvalidScope
+	}
+
+	if release.OnClassic {
+		q.Set("confinement", "strict,classic")
+	} else {
+		q.Set("confinement", "strict")
+	}
+
+	u := s.endpointURL(findEndpPath, q)
+	reqOptions := &requestOptions{
+		Method:   "GET",
+		URL:      u,
+		Accept:   jsonContentType,
+		APILevel: apiV2Endps,
+	}
+
+	var searchData searchV2Results
+
+	// TODO: use retryRequestDecodeJSON (may require content-type check there,
+	// requires checking other handlers, their tests and store).
+	doRequest := func() (*http.Response, error) {
+		return s.doRequest(ctx, s.client, reqOptions, user)
+	}
+	readResponse := func(resp *http.Response) error {
+		ok := (resp.StatusCode == 200 || resp.StatusCode == 201)
+		ct := resp.Header.Get("Content-Type")
+		// always decode on success; decode failures only if body is not empty
+		if !ok && (resp.ContentLength == 0 || ct != jsonContentType) {
+			return nil
+		}
+		return json.NewDecoder(resp.Body).Decode(&searchData)
+	}
+	resp, err := httputil.RetryRequest(u.String(), doRequest, readResponse, defaultRetryStrategy)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != 200 {
+		// fallback to search v1; v2 may not be available on some proxies
+		if resp.StatusCode == 404 {
+			verstr := resp.Header.Get("Snap-Store-Version")
+			ver, err := strconv.Atoi(verstr)
+			if err != nil {
+				logger.Debugf("Bogus Snap-Store-Version header %q.", verstr)
+			} else if ver < 20 {
+				return s.findV1(ctx, search, user)
+			}
+		}
+		if len(searchData.ErrorList) > 0 {
+			return nil, translateSnapActionError("", "", searchData.ErrorList[0].Code, searchData.ErrorList[0].Message, nil)
```

Comment:
(nitpick) Hm, we check if the len of error list if bigger than 1 but we only translate error [0] - should we be more defensive and warn about ErrorLists with size > 1 ?

Reply:
If you mean logging than yes, sounds like a good idea. We don't expect more than one error here in search context, it's just how the json protocol is defined in store for multi-snap ops.